akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -984,31 +997,176 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
   /**
    * Obtain a callback object that can be used asynchronously to re-enter the
    * current [[GraphStage]] with an asynchronous notification. The [[invoke()]] method of the returned
-   * [[AsyncCallback]] is safe to be called from other threads and it will in the background thread-safely
-   * delegate to the passed callback function. I.e. [[invoke()]] will be called by the external world and
+   * [[AsyncCallback]] is safe to be called from other threads. It will in the background thread-safely
+   * delegate to the passed callback function. I.e. [[invoke()]] will be called by other thread and
    * the passed handler will be invoked eventually in a thread-safe way by the execution environment.
    *
+   * In case stream is not yet materialized [[AsyncCallback]] will buffer events until stream is available.
+   *
+   * [[AsyncCallback.invokeWithFeedback()]] has an internal promise that will be failed if event cannot be processed
+   * due to stream completion.
+   *
    * This object can be cached and reused within the same [[GraphStageLogic]].
    */
   final def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
-    new AsyncCallback[T] {
-      override def invoke(event: T): Unit =
-        interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
+    val result = new ConcurrentAsyncCallback[T](handler)
+    concurrentAsyncCallback.add(result)
```

Comment:
I'd like to understand the performance overhead compared to the old implementation. At a few places internally we use async callbacks and the performance might matter, e.g. in the Artery `SendQueue`. I only care about "steady-state", not initial calls during initialization.

Reply:
for "steady-state" and without feedback promise - it's just checking for a state at the beginning and then send message to an actor as it was before.